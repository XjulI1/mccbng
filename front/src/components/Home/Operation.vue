<template>
  <div class="operation">
    <input
      :id="checkBoxIDValue"
      v-model="modelValue"
      type="checkbox"
      @change="updateCheckOp"
    />
    <div class="label" :class="css.category">
      <label :for="checkBoxIDValue" :data-id="operation.IDop">
        {{ operation.NomOp }}
        <br />
        {{ dateOperation }}
      </label>
    </div>
    <div class="montant" :class="css.montant">
      <Currency :amount="operation.MontantOp" />
    </div>
    <div>
      <router-link
        :to="'/editOperation/' + operation.IDop"
        class="edit-operation"
      >
        ...
      </router-link>
    </div>
  </div>
</template>

<script setup>
import { computed, ref, watch } from "vue";
import { useStore } from "vuex";
import {
  checkBoxID,
  generateCssVariables,
  generateDateOperationVariables,
} from "@/helpers/components/Operation";
import Currency from "../Currency";

const props = defineProps({
  operation: {
    type: Object,
    default: () => {},
  },
});

const store = useStore();

const dateOperation = ref(generateDateOperationVariables(props.operation));
const css = ref(generateCssVariables(props.operation));
const checkBoxIDValue = checkBoxID(props.operation.IDop);

const modelValue = computed(() => {
  return props.operation.CheckOp;
});

watch(
  () => props.operation,
  () => {
    dateOperation.value = generateDateOperationVariables(props.operation);
    css.value = generateCssVariables(props.operation);
  }
);

const updateCheckOp = () => {
  store.dispatch("updateOperation", {
    ...props.operation,
    CheckOp: !props.operation.CheckOp,
  });
};
</script>

<style lang="scss" scoped>
.operation {
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 60px;
  padding-left: 10px;
  padding-right: 10px;

  @media all and (min-width: $desktop_BP_min_width) {
    max-width: 600px;
  }

  @media screen and (max-width: $mobile_BP_max_width) {
    max-width: 768px;
    height: 80px;
  }

  input[type="checkbox"] {
    margin-right: 10px;
  }

  .label {
    flex-grow: 2;
  }

  .montant {
    min-width: 90px;
    padding-right: 15px;
    text-align: right;

    &.montantIn {
      color: green;
    }

    &.montantOut {
      color: red;
    }
  }

  .noCategory {
    color: grey;
  }

  .edit-operation {
    text-decoration: none;
  }
}
</style>
