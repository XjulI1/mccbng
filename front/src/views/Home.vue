<template>
  <div class="home">
    <router-view />
    <OperationList :OperationRenderer="OperationRenderer" />
  </div>
</template>

<script setup>
  import OperationList from '../components/OperationList.vue'
  import OperationRenderer from '../components/Home/Operation.vue'
</script>

<style lang="scss" scoped>
@media screen and (max-width: 767px) {
  .home {
    margin-bottom: $navbar-height;
  }
}

@media screen and (min-width: 768px) {
  .home {
    display: flex;
  }
}
</style>
