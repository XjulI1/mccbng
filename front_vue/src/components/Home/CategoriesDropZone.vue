<template>
  <div
          class="categories-drop-zone"
          :class="{'display-zone' : displayCategoriesDropZone}"
  >
    <div class="category-container">
      <draggable
              v-for="category in categoryList "
              :key="'category-' + category.IDcat"
              class="category"
              :data-id="category.IDcat"
              :group="{name: 'category', put: ['operation']}"
              @add="addNewOperation"
      >
        <div :class="{'actual-drag-cat' : displayActualDragCat === category.IDcat}">
          <b>{{ category.Nom }}</b>
        </div>
      </draggable>
    </div>
  </div>
</template>

<script>
  import { mapGetters, mapState } from 'vuex'
  import draggable from 'vuedraggable'

  export default {
    name: 'CategoriesDropZone',

    components: { draggable },

    computed: {
      ...mapState({
        userID: state => state.user.id,
        categoryList: state => state.category.list,
        displayCategoriesDropZone: state => state.display.categories_drop_zone,
        displayActualDragCat: state => state.display.actual_drag_cat
      }),
      ...mapGetters(['operationFromCurrentList'])
    },

    watch: {
      userID () {
        this.$store.dispatch('fetchCategoryList')
      }
    },

    methods: {
      addNewOperation (event) {
        const categoryID = event.target.dataset.id
        const operationID = event.item.dataset.id

        const operation = this.operationFromCurrentList(operationID)

        operation.IDcat = categoryID

        this.$store.dispatch('updateOperation', operation)

        event.item.remove()
      }
    }
  }
</script>

<style lang="scss" scoped>
  .display-zone {
    display: block !important;
  }

  .actual-drag-cat {
    color: orange;
  }

  .categories-drop-zone {
    position: fixed;
    z-index: 1000;
    bottom: 0;
    left: 0;
    width: 100%;
    display: none;
    background-color: lightgrey;
  }

  .category-container {
    display: flex;
    flex-wrap: wrap;
    text-align: center
  }

  .category {
    padding: 1rem;
    width: 25%;
    border: darkgrey 1px solid;
  }

  @media screen and (max-width: 767px) {
    .category {
      width: 33%;
    }
  }

  @media screen and (min-width: 768px) {
    .category {
      width: 25%;
    }
  }

  .hover-drag {
    background-color: darkgrey;
  }
</style>
