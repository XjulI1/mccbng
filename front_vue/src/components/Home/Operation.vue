<template>
  <div class="operation container">
    <div class="row">
      <div class="col-1">
        <input type="checkbox" :id="checkBoxID" v-model="operation.CheckOp" @change="updateCheckOp"/>
      </div>
      <draggable class="col-7"
                 :class="css.category"
                 @start="startDrag"
                 @end="endDrag"
                 :group="{name: 'operation', pull: 'clone', put: ['false'] }"
                 :data-idcat="operation.IDcat">
        <label :for="checkBoxID" :data-id="operation.IDop">
          {{operation.NomOp}}
          <br>
          {{dateOperation}}
        </label>
      </draggable>
      <div class="col-3 center-text" :class="css.montant">
        {{operation.MontantOp.toLocaleString()}}{{$store.state.currency}}
      </div>
      <div class="col-1">
        <router-link :to="'/editOperation/' + operation.IDop" class="edit-operation">...</router-link>
      </div>
    </div>
  </div>
</template>

<script>
  import draggable from 'vuedraggable'

  export default {
    name: 'Operation',
    props: ['operation'],

    components: { draggable },

    watch: {
      operation () {
        this.dateOperation = this.generateDataOperationVariables()
        this.css = this.generateCssVariables()
      }
    },

    data () {
      return {
        dateOperation: this.generateDataOperationVariables(),
        css: this.generateCssVariables(),
        checkBoxID: 'checkboxOperationID-' + this.operation.IDop
      }
    },

    methods: {
      generateCssVariables () {
        return {
          category: this.operation.IDcat === 0 ? 'noCategory' : '',
          montant: this.operation.MontantOp > 0 ? 'montantIn' : 'montantOut'
        }
      },

      generateDataOperationVariables () {
        return new Date(this.operation.DateOp).toLocaleDateString()
      },

      updateCheckOp () {
        this.$store.dispatch('updateOperation', this.operation)
      },

      startDrag (event) {
        this.$store.dispatch('toggleCategoriesDropZone')
        this.$store.dispatch('actualDragCategory', event.srcElement.dataset.idcat)
      },

      endDrag () {
        this.$store.dispatch('toggleCategoriesDropZone')
      }
    }
  }
</script>

<style scoped>
  input[type=checkbox] {
    margin-top: 1.0rem;
  }

  .center-text {
    text-align: center;
  }

  .montantIn {
    color: green;
  }

  .montantOut {
    color: red
  }

  .noCategory {
    color: grey;
  }

  .edit-operation {
    text-decoration: none;
  }
</style>
