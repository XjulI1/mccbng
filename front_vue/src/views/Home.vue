<template>
  <div class="home">
    <router-view/>
    <operation-list/>
  </div>
</template>

<script>
  import OperationList from '../components/OperationList'

  export default {
    name: 'Home',
    components: { OperationList }
  }
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style lang="scss" scoped>
  @import '@/global.scss';

  @media screen and (max-width: 767px) {
    .home {
      margin-bottom: $navbar-height;
    }
  }

  @media screen and (min-width: 768px) {
    .home {
      display: flex;
    }
  }
</style>
