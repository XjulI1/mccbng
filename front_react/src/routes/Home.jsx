import React from 'react'

import './styles/Home.scss'

class Home extends React.Component {
  render () {
    return <div className="home-page">
      Home page
    </div>
  }
}

export default Home
