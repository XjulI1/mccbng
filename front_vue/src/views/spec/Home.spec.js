import { createLocalVue, shallowMount } from '@vue/test-utils'

import Router from 'vue-router'

import Home from '@/views/Home'

const localVue = createLocalVue()

localVue.use(Router)

describe('Home.vue', () => {
  test('should match snapshot', () => {
    const wrapper = shallowMount(Home, {
      localVue
    })

    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })
})
