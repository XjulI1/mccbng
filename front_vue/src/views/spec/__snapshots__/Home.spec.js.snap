// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue should match snapshot 1`] = `
<div
  class="home"
>
  <routerview-stub
    name="default"
  />
   
  <operationlist-stub />
</div>
`;
